File: frontend/js/services/gemini-service.js
// Prism AI - Gemini Service with Python Backend Integration
// Handles communication with the Python backend for AI recommendations

// Imports removed - using global variables
// const { APPWRITE_CONFIG, authService } = window;
// const { preferencesService } = window;

class PrismGeminiService {
    constructor() {
        this.pythonApiUrl = window.APPWRITE_CONFIG.pythonApiUrl;
        this.isProcessing = false;
        this.requestQueue = [];
        this.maxRetries = 3;
        // Check if backend is healthy
        if (!this.backendHealthy) {
            console.warn('🔄 Backend unhealthy, attempting health check...');
            const healthCheck = await this.checkBackendHealth();
            if (!healthCheck.healthy) {
                return this.getFallbackResponse(userQuery, userProfile, 'Backend unavailable');
            }
        }

        // Check rate limiting
        if (!this.checkRateLimit()) {
            throw new Error('Rate limit exceeded. Please wait a moment before trying again.');
        }

        const maxRetries = this.maxRetries;
        let attempt = 0;

        while (attempt < maxRetries) {
            try {
                console.log('🤖 Calling Gemini AI via Python Backend...');
                console.log('📡 Backend URL:', this.apiUrl);

                // Build complete user profile with defaults
                const mergedProfile = await this.buildUserProfile(userProfile);

                const payload = {
                    userQuery: userQuery || 'Hello',
                    userProfile: mergedProfile,
                    imageData: imageData || null
                };

                console.log('📤 Sending to Python backend:', {
                    userQuery: payload.userQuery,
                    profileKeys: Object.keys(payload.userProfile),
                    hasImage: !!payload.imageData
                });

                // Call Python Backend with timeout
                const controller = new AbortController();
                const timeoutId = setTimeout(() => controller.abort(), this.timeout);

                const response = await fetch(this.apiUrl, {
                    method: 'POST',
                    headers: {
                        'Content-Type': 'application/json',
                    },
                    body: JSON.stringify(payload),
                    signal: controller.signal
                });

                clearTimeout(timeoutId);
                console.log('📡 Backend response status:', response.status);

                if (!response.ok) {
                    const errorData = await response.json().catch(() => ({}));
                    throw new Error(errorData.error || `Backend error: ${response.status}`);
                }

                const result = await response.json();
                console.log('✅ Python backend response received');

                // CHECK IF BACKEND RETURNED FALLBACK DATA
                if (result.fallback === true) {
                    console.warn('⚠️ BACKEND RETURNED FALLBACK - GEMINI API FAILED!');
                    console.warn('   This means the Gemini API call failed on the backend.');
                    console.warn('   Check the backend terminal logs for the error details!');
                    console.warn('   Backend response:', result);
                    if (result.backendError) {
                        console.warn('   backendError:', result.backendError);
                    }
                    if (result.backendDetails) {
                        console.warn('   backendDetails:', result.backendDetails);
                    }
                    if (result.modelErrors) {
                        console.warn('   modelErrors:', result.modelErrors);
                    }
                }

                // Track successful request
                this.trackRequest();

                // Ensure proper response format
                if (result.success && result.data) {
                    return {
                        success: true,
                        generalAdvice: result.data.generalAdvice || 'Here are some recommendations for you:',
                        products: result.data.products || [],
                        additionalTips: result.data.additionalTips || [],
                        processingTime: result.processingTime,
                        timestamp: result.timestamp,
                        isFallback: result.fallback === true  // Flag fallback responses
                    };
                } else {
                    throw new Error(result.error || 'Invalid response format from backend');
                }

            } catch (error) {
                attempt++;
                console.error(`❌ Backend attempt ${attempt}/${maxRetries} failed:`, error.message);

                // Check for specific error types
                if (error.name === 'AbortError') {
                    console.error('⏰ Request timeout - backend taking too long');
                } else if (error.message.includes('fetch') || error.name === 'TypeError') {
                    console.error('🚨 Cannot connect to Python backend!');
                    console.error('💡 Make sure backend is running: cd backend/python-api && python app.py');
                    this.backendHealthy = false;
                }

                if (attempt >= maxRetries) {
                    // Return structured fallback response
                    return this.getFallbackResponse(userQuery, userProfile, error.message);
                }

                // Exponential backoff
                await this.delay(1000 * Math.pow(2, attempt - 1));
            }
        }
    }

    /**
     * Legacy method for compatibility with existing code
     */
    async getRecommendations(userQuery, userProfile, imageData = null) {
        return this.generateResponse(userQuery, userProfile, imageData);
    }


    /**
     * Test the backend with a simple request
     */
    async testBackend() {
        try {
            const testUrl = `${this.pythonBackend.baseUrl}${this.pythonBackend.endpoints.test}`;
            const response = await fetch(testUrl, {
                method: 'POST',
                headers: {
                    'Content-Type': 'application/json',
                },
                body: JSON.stringify({
                    test: true,
                    message: 'Frontend test request'
                }),
                signal: AbortSignal.timeout(5000)
            });

            if (!response.ok) {
                throw new Error(`Test failed: ${response.status}`);
            }

            const data = await response.json();
            console.log('🧪 Backend test successful:', data);
            return { success: true, data };
        } catch (error) {
            console.error('❌ Backend test failed:', error);
            return { success: false, error: error.message };
        }
    }

    /**
     * Get fallback response when API fails
     */
    getFallbackResponse(userQuery, preferences, errorMessage = 'Technical difficulties') {
        const fallbackProducts = this.generateFallbackProducts(preferences);

        let advice = `I'm experiencing some technical difficulties, but here are some general recommendations based on your profile.`;

        // Add specific error context
        if (errorMessage.includes('Backend unavailable') || errorMessage.includes('fetch')) {
            advice = `The AI backend is currently unavailable. Here are some general recommendations while we reconnect.`;
        } else if (errorMessage.includes('timeout')) {
            advice = `The AI is taking longer than usual to respond. Here are some quick recommendations.`;
        }

        return {
            success: true,
            generalAdvice: advice,
            products: fallbackProducts,
            additionalTips: [
                "Always patch test new products before full application",
                "Stay hydrated and maintain a consistent skincare routine",
                "Consult with a dermatologist for persistent concerns",
                "Try asking your question again in a moment"
            ],
            fallback: true,
            error: errorMessage,
            processingTime: "0.1s",
            timestamp: new Date().toISOString()
        };
    }

    /**
     * Generate fallback product recommendations
     */
    generateFallbackProducts(preferences) {
        const skinType = preferences?.skinType || 'normal';
        const budget = preferences?.skincareBudget || 'mid';

        const productDatabase = {
            oily: [
                {
                    name: "Minimalist 10% Niacinamide Face Serum (30ml)",
                    brand: "Minimalist",
                    category: "Skincare",
                    price: "₹599",
                    ingredients: ["Niacinamide", "Zinc"],
                    purchaseLink: "https://www.amazon.in/dp/B08GY59SH4",
                    compatibility: 90,
                    pros: ["Controls oil production", "Minimizes pores", "Budget-friendly"],
                    cons: ["Can be drying if overused"],
                    applicationTips: "Apply a few drops after cleansing, before moisturizer"
                },
                {
                    name: "Minimalist 2% Salicylic Acid Serum (30ml)",
                    brand: "Minimalist",
                    category: "Skincare",
                    price: "₹599",
                    ingredients: ["Salicylic Acid", "LHA"],
                    purchaseLink: "https://www.amazon.in/dp/B08L4S8F3Z",
                    compatibility: 88,
                    pros: ["Unclogs pores", "Reduces blackheads", "Controls excess oil"],
                    cons: ["May be drying initially"],
                    applicationTips: "Use 2-3 times per week in the evening"
                },
                {
                    name: "Plum Green Tea Renewed Clarity Face Wash (75ml)",
                    brand: "Plum",
                    category: "Skincare",
                    price: "₹345",
                    ingredients: ["Green Tea", "Glycolic Acid"],
                    purchaseLink: "https://www.amazon.in/dp/B01MSJXZ7L",
                    compatibility: 86,
                    pros: ["Oil control", "Gentle exfoliation", "Natural ingredients"],
                    cons: ["May dry out skin if used twice daily"],
                    applicationTips: "Use once daily, preferably in the morning"
                }
            ],
            dry: [
                {
                    name: "Cetaphil Moisturizing Cream (550ml)",
                    brand: "Cetaphil",
                    category: "Skincare",
                    price: "₹999",
                    ingredients: ["Glycerin", "Panthenol", "Sweet Almond Oil"],
                    purchaseLink: "https://www.amazon.in/dp/B003MJG19K",
                    compatibility: 92,
                    pros: ["Deep hydration", "Dermatologist recommended", "Non-greasy"],
                    cons: ["Large bottle may be inconvenient for travel"],
                    applicationTips: "Apply to clean, damp skin twice daily for best results"
                },
                {
                    name: "Minimalist Sepicalm 3% + Oat Moisturizer (100ml)",
                    brand: "Minimalist",
                    category: "Skincare",
                    price: "₹699",
                    ingredients: ["Sepicalm", "Oat Extract", "Ceramides"],
                    purchaseLink: "https://www.amazon.in/dp/B08T6X916Q",
                    compatibility: 90,
                    pros: ["Soothes dry skin", "Strengthens skin barrier", "Lightweight texture"],
                    cons: ["Smaller bottle size"],
                    applicationTips: "Apply morning and night after serum"
                }
            ],
            combination: [
                {
                    name: "Neutrogena Hydro Boost Water Gel (50g)",
                    brand: "Neutrogena",
                    category: "Skincare",
                    price: "₹899",
                    ingredients: ["Hyaluronic Acid", "Glycerin"],
                    purchaseLink: "https://www.nykaa.com/neutrogena-hydro-boost-water-gel/p/260639",
                    compatibility: 89,
                    pros: ["Lightweight hydration", "Non-greasy", "Oil-free formula"],
                    cons: ["Small jar size"],
                    applicationTips: "Apply as last step in routine before sunscreen"
                },
                {
                    name: "Minimalist 10% Niacinamide Face Serum (30ml)",
                    brand: "Minimalist",
                    category: "Skincare",
                    price: "₹599",
                    ingredients: ["Niacinamide", "Zinc"],
                    purchaseLink: "https://www.amazon.in/dp/B08GY59SH4",
                    compatibility: 87,
                    pros: ["Balances oil production", "Minimizes pores", "Budget-friendly"],
                    cons: ["May cause purging initially"],
                    applicationTips: "Use twice daily after cleansing"
                }
            ],
            sensitive: [
                {
                    name: "Simple Kind to Skin Hydrating Light Moisturiser (125ml)",
                    brand: "Simple",
                    category: "Skincare",
                    price: "₹385",
                    ingredients: ["Vitamin E", "Vitamin B5", "Bisabolol"],
                    purchaseLink: "https://www.nykaa.com/simple-kind-to-skin-hydrating-light-moisturiser/p/129162",
                    compatibility: 92,
                    pros: ["Fragrance-free", "Lightweight", "Sensitive skin friendly"],
                    cons: ["Not enough for very dry skin"],
                    applicationTips: "Use morning and evening after cleansing"
                }
            ],
            normal: [
                {
                    name: "Neutrogena Hydro Boost Water Gel (50g)",
                    brand: "Neutrogena",
                    category: "Skincare",
                    price: "₹899",
                    ingredients: ["Hyaluronic Acid", "Glycerin"],
                    purchaseLink: "https://www.nykaa.com/neutrogena-hydro-boost-water-gel/p/260639",
                    compatibility: 87,
                    pros: ["Lightweight hydration", "Non-greasy"],
                    cons: ["Jar may require frequent restock"],
                    applicationTips: "Apply as last step in routine (before sunscreen in AM)"
                }
            ]
        };

        return productDatabase[skinType] || productDatabase.normal;
    }

    /**
     * Rate limiting
     */
    checkRateLimit() {
        const now = Date.now();

        // Remove timestamps older than window
        this.requestTimestamps = this.requestTimestamps.filter(
            timestamp => now - timestamp < this.requestWindow
        );

        // Check if under limit
        return this.requestTimestamps.length < this.maxRequestsPerMinute;
    }

    trackRequest() {
        this.requestTimestamps.push(Date.now());
    }

    /**
     * Utility functions
     */
    formatBudget(budget) {
        const budgetMap = {
            'budget': 'Under ₹500',
            'mid': '₹500-₹1500',
            'premium': '₹1500-₹3000',
            'luxury': '₹3000+'
        };
        return budgetMap[budget] || 'Any';
    }

    formatAgeRange(range) {
        const rangeMap = {
            'teens': '13-19',
            'twenties': '20-29',
            'thirties': '30-39',
            'forties': '40-49',
            'fifties': '50-59',
            'sixty-plus': '60+'
        };
        return rangeMap[range] || 'Not specified';
    }

    delay(ms) {
        return new Promise(resolve => setTimeout(resolve, ms));
    }

    /**
     * Get estimated wait time for rate limit
     */
    getWaitTime() {
        if (this.requestTimestamps.length === 0) return 0;

        const oldestTimestamp = this.requestTimestamps[0];
        const waitTime = this.requestWindow - (Date.now() - oldestTimestamp);

        return Math.max(0, waitTime);
    }
}

// Export singleton instance
export const geminiService = new GeminiService();

File: backend/python-api/services/gemini_service.py
"""
Gemini AI Service
Handles all interactions with Google's Gemini AI API
"""

import os
import requests
import json
import logging
from typing import Dict, Any, Optional

logger = logging.getLogger(__name__)

class GeminiService:
    def __init__(self):
        # Load Gemini API key from environment variables only. No hard-coded defaults in code.
        self.api_key = os.getenv('GEMINI_API_KEY', '').strip()
        if not self.api_key:
            logger.error("GEMINI_API_KEY is not set in environment variables. please set GEMINI_API_KEY in your shell or deployment environment")
        
        self.model = os.getenv('GEMINI_MODEL', 'gemini-1.5-flash')
        self.max_tokens = int(os.getenv('MAX_TOKENS', '2048'))
        self.temperature = float(os.getenv('TEMPERATURE', '0.7'))
        self.timeout = int(os.getenv('REQUEST_TIMEOUT', '60'))  # Increased for Google Search grounding
        
        # Updated model fallbacks with NEW Gemini 2.0+ models (verified working as of Nov 2024)
        self.model_fallbacks = [
            'gemini-2.5-flash',      # Fastest, recommended for production
            'gemini-2.0-flash',      # Backup option
            'gemini-2.5-pro'         # Most capable, if the others fail
        ]
        
        if not self.api_key:
            raise ValueError("GEMINI_API_KEY is required. Set GEMINI_API_KEY environment variable in your shell or deployment platform.")
    
    def get_recommendations(self, user_query: str, user_profile: Dict[str, Any], image_data: Optional[str] = None) -> Dict[str, Any]:
        """
        Get AI recommendations from Gemini API
        
        Args:
            user_query: User's question/request
            user_profile: User's preferences and profile data
            image_data: Optional base64 encoded image
            
        Returns:
            Dict with success status and data/error
        """
        logger.info("="*60)
        logger.info("🎯 Starting get_recommendations")
        logger.info(f"   Query: {user_query[:100]}...")
        logger.info(f"   Profile keys: {list(user_profile.keys()) if user_profile else 'None'}")
        logger.info(f"   Has image: {bool(image_data)}")
        logger.info("="*60)
        
        try:
            model_errors = []
            # Build the prompt
            logger.info("📝 Building prompt...")
            prompt = self._build_prompt(user_query, user_profile, image_data)
            logger.info(f"✅ Prompt built: {len(prompt)} characters")
            
            # Try each model until one works
            logger.info(f"🔄 Trying {len(self.model_fallbacks)} models: {self.model_fallbacks}")
            for model in self.model_fallbacks:
                try:
                    logger.info(f"Trying Gemini model: {model}")
                    
                    response = self._call_gemini_api(prompt, model)
                    
                    if response:
                        logger.info(f"Successfully got response from {model}")
                        return {
                            'success': True,
                            'data': response,
                            'model_used': model
                        }
                        
                except Exception as e:
                    err_msg = str(e)
                    logger.error(f"❌ Model {model} failed: {err_msg}")
                    logger.error(f"   Full error details: {repr(e)}")
                    model_errors.append({'model': model, 'error': err_msg})
                    continue
            
            # If all models failed
            return {
                'success': False,
                'error': 'All Gemini models failed. Please try again later.',
                'details': 'No available models could process the request',
                'errors': model_errors
            }
            
        except Exception as e:
            logger.error(f"Error in get_recommendations: {str(e)}")
            return {
                'success': False,
                'error': str(e)
            }
    
    def _call_gemini_api(self, prompt: str, model: str) -> Optional[str]:
        """
        Make API call to Gemini
        
        Args:
            prompt: The formatted prompt
            model: Model name to use
            
        Returns:
            AI response text or None if failed
        """
        # Use v1 for stability
        api_version = 'v1'
        url = f"https://generativelanguage.googleapis.com/{api_version}/models/{model}:generateContent?key={self.api_key}"
        
        # Log the API call details
        logger.info(f"🚀 Making Gemini API call with India-specific prompting:")
        logger.info(f"   Model: {model}")
        logger.info(f"   URL: {url[:80]}...")
        logger.info(f"   API Key (first 10 chars): {self.api_key[:10]}...")
        logger.info(f"   Prompt length: {len(prompt)} characters")
        logger.info(f"   🇮🇳 Validation: Non-Indian links will be filtered out")
        
        payload = {
            "contents": [
                {
                    "role": "user",
                    "parts": [{"text": prompt}]
                }
            ],
            "generationConfig": {
                "temperature": self.temperature,
                "topK": 20,
                "topP": 0.9,
                "maxOutputTokens": self.max_tokens
            },
        }
        
        headers = {
            "Content-Type": "application/json"
        }
        
        logger.info("📡 Sending POST request to Gemini API...")
        
        try:
            response = requests.post(
                url,
                json=payload,
                headers=headers,
                timeout=self.timeout
            )
            
            logger.info(f"📥 Received response: Status {response.status_code}")
        except requests.exceptions.Timeout:
            logger.error(f"⏱️ Request timeout after {self.timeout} seconds")
            raise Exception(f"Request timeout after {self.timeout} seconds")
        except requests.exceptions.ConnectionError as e:
            logger.error(f"🔌 Connection error: {str(e)}")
            raise Exception(f"Connection error: {str(e)}")
        except Exception as e:
            logger.error(f"❌ Request failed: {str(e)}")
            raise
        
        if response.status_code == 200:
            data = response.json()
            logger.info(f"✅ API call successful! Response keys: {list(data.keys())}")
            
            if 'candidates' in data and len(data['candidates']) > 0:
                content = data['candidates'][0]['content']['parts'][0]['text']
                logger.info(f"📝 Generated content length: {len(content)} characters")
                return content
            else:
                logger.error(f"❌ No candidates in response. Full response: {json.dumps(data, indent=2)}")
                raise Exception("No candidates in response")
        
        elif response.status_code == 404:
            logger.error(f"❌ Model not found: {model}")
            raise Exception(f"Model {model} not found or not supported")
        
        else:
            error_data = response.json() if response.content else {}
            error_msg = error_data.get('error', {}).get('message', f"HTTP {response.status_code}")
            logger.error(f"❌ API error {response.status_code}: {error_msg}")
            logger.error(f"   Full error response: {json.dumps(error_data, indent=2)}")
            raise Exception(f"API error: {error_msg}")
    
    def _build_prompt(self, user_query: str, user_profile: Dict[str, Any], image_data: Optional[str] = None) -> str:
        """
        Build optimized prompt for Gemini AI
        
        Args:
            user_query: User's question
            user_profile: User preferences
            image_data: Optional image data
            
        Returns:
            Formatted prompt string
        """
        # Format user profile
        profile_context = self._format_user_profile(user_profile)
        
        prompt = f"""You are an expert beauty and fashion consultant specializing in the INDIAN MARKET with deep knowledge of skincare, haircare, and fashion products available in India.

━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
USER PROFILE:
━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
{profile_context}
━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━

USER QUESTION: {user_query}

🚨 CRITICAL INSTRUCTION - READ CAREFULLY:
═══════════════════════════════════════════════════════════════
🇮🇳 RECOMMEND ONLY FROM THE VERIFIED PRODUCT DATABASE BELOW

YOU MUST:
1. ✅ Choose products ONLY from the "VERIFIED INDIAN PRODUCTS" list below
2. ✅ Use EXACT product names and links as provided
3. ✅ Use EXACT prices in ₹ (Indian Rupees) as listed
4. ✅ Match products to user's profile (skin type, hair type, style)
5. ✅ Provide 3-5 products that best match user's needs
6. ❌ DO NOT make up new products not in the list
7. ❌ DO NOT modify product names or links
8. ❌ DO NOT use amazon.com or any non-Indian sites
9. ❌ DO NOT provide prices in dollars ($)
10. ❌ DO NOT create generic/fake product links

LINK FORMAT REQUIREMENTS:
✓ CORRECT: "https://www.amazon.in/Cetaphil-Gentle-Cleanser-Face-125ml/dp/B07PYZFMMX"
✓ CORRECT: "https://www.nykaa.com/cetaphil-gentle-skin-cleanser/p/22317"
✗ WRONG: "https://www.amazon.in/s?k=moisturizer" (search URL)
✗ WRONG: "https://www.amazon.in/product-name" (generic/fake URL)

━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
VERIFIED INDIAN PRODUCTS DATABASE (Choose ONLY from here)
━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━

🧴 MOISTURIZERS:
─────────────────────────────────────────────────────────────
1. Cetaphil Moisturizing Cream (550ml)
   Price: ₹999 | Skin: Dry, Sensitive
   Link: https://www.amazon.in/dp/B003MJG19K
   Ingredients: Glycerin, Panthenol, Sweet Almond Oil

2. Minimalist Sepicalm 3% + Oat Moisturizer (100ml)
   Price: ₹399 | Skin: All types, Sensitive
   Link: https://www.amazon.in/dp/B08T6X916Q
   Ingredients: Sepicalm, Oat Extract, Squalane

3. Plum Hello Aloe Just Gel (150ml)
   Price: ₹299 | Skin: Oily, Combination
   Link: https://www.amazon.in/dp/B076PWFVB1
   Ingredients: Aloe Vera, Glycerin

4. Neutrogena Hydro Boost Water Gel (50g)
   Price: ₹899 | Skin: All types
   Link: https://www.amazon.in/s?k=Neutrogena+Hydro+Boost+Water+Gel
   Ingredients: Hyaluronic Acid, Glycerin

5. Simple Kind to Skin Hydrating Light Moisturiser (125ml)
   Price: ₹385 | Skin: Dry, Sensitive
   Link: https://www.nykaa.com/simple-kind-to-skin-hydrating-light-moisturiser/p/129162
   Ingredients: Vitamin E, Vitamin B5, Bisabolol

🧼 CLEANSERS:
─────────────────────────────────────────────────────────────
6. Cetaphil Gentle Skin Cleanser (125ml)
   Price: ₹525 | Skin: All types, Sensitive
   Link: https://www.amazon.in/dp/B07PYZFMMX
   Ingredients: Glycerin, Panthenol, Niacinamide

7. Plum Green Tea Renewed Clarity Face Wash (75ml)
   Price: ₹299 | Skin: Oily, Acne-prone
   Link: https://www.amazon.in/dp/B01MSJXZ7L
   Ingredients: Green Tea, Glycolic Acid

8. Simple Refreshing Facial Wash Gel (150ml)
   Price: ₹350 | Skin: All types
   Link: https://www.amazon.in/dp/B00COMJSR8
   Ingredients: Vitamin E, Vitamin B5, Chamomile

💧 SERUMS:
─────────────────────────────────────────────────────────────
9. Minimalist 10% Niacinamide Face Serum (30ml)
   Price: ₹599 | Skin: Oily, Acne-prone, Large pores
   Link: https://www.amazon.in/dp/B08GY59SH4
   Ingredients: 10% Niacinamide, 1% Zinc

10. Minimalist 2% Salicylic Acid Serum (30ml)
    Price: ₹599 | Skin: Oily, Acne-prone
    Link: https://www.amazon.in/dp/B08L4S8F3Z
    Ingredients: 2% Salicylic Acid, LHA

11. Minimalist 0.3% Retinol Serum (30ml)
    Price: ₹699 | Skin: Aging, Dull
    Link: https://www.amazon.in/dp/B08GXWH3PJ
    Ingredients: 0.3% Retinol, Squalane

🧴 SUNSCREEN:
─────────────────────────────────────────────────────────────
12. Minimalist SPF 50 Sunscreen (50ml)
    Price: ₹449 | Skin: All types
    Link: https://www.amazon.in/dp/B08X6JYB9H
    Ingredients: SPF 50 PA++++, Multi-Vitamin

13. Neutrogena Ultra Sheer Dry Touch Sunblock SPF 50+ (30ml)
    Price: ₹499 | Skin: Oily, Combination
    Link: https://www.amazon.in/dp/B073DSCT4V
    Ingredients: Helioplex, Oil-free

💆 HAIRCARE:
─────────────────────────────────────────────────────────────
14. L'Oreal Paris 6 Oil Nourish Shampoo (704ml)
    Price: ₹799 | Hair: Dry, Damaged
    Link: https://www.amazon.in/dp/B08B3WRRPK
    Ingredients: 6 Micro Oils

15. Tresemme Keratin Smooth Shampoo (580ml)
    Price: ₹525 | Hair: Frizzy, Dry
    Link: https://www.amazon.in/dp/B01NAPMM6N
    Ingredients: Keratin, Argan Oil

16. Pantene Advanced Hairfall Solution Shampoo (650ml)
    Price: ₹499 | Hair: Hair fall
    Link: https://www.amazon.in/dp/B08BR25C19
    Ingredients: Pro-Vitamin B5, Fermented Rice Water

17. WOW Skin Science Onion Black Seed Oil Shampoo (300ml)
    Price: ₹449 | Hair: Hair fall, Thinning
    Link: https://www.amazon.in/dp/B07V3J8KS4
    Ingredients: Red Onion, Black Seed Oil

👔 FASHION (MEN):
─────────────────────────────────────────────────────────────
18. Allen Solly Men's Regular Fit Shirt
    Price: ₹1,299 | Style: Formal, Business
    Link: https://www.amazon.in/dp/B0BY4HTPQ1
    Details: Cotton blend, Multiple colors

19. Van Heusen Men's Casual Shirt
    Price: ₹1,099 | Style: Casual, Smart casual
    Link: https://www.amazon.in/dp/B0CPFH3B99
    Details: Cotton, Checkered patterns

👗 FASHION (WOMEN):
─────────────────────────────────────────────────────────────
20. Only Women's Regular Fit Top
    Price: ₹999 | Style: Casual, Western
    Link: https://www.amazon.in/dp/B0CXV7HX8Y
    Details: Cotton blend, Various colors

21. Vero Moda Women's Casual Dress
    Price: ₹1,599 | Style: Party, Casual
    Link: https://www.amazon.in/dp/B0B8MQWXNJ
    Details: Polyester, A-line fit

━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
SELECTION RULES:
• For DRY skin → Recommend products 1, 2, 5, 6, 8
• For OILY skin → Recommend products 3, 7, 9, 10, 13
• For SENSITIVE skin → Recommend products 1, 2, 5, 6, 8
• For ACNE-PRONE skin → Recommend products 7, 9, 10
• For DRY hair → Recommend products 14, 15
• For FRIZZY hair → Recommend products 15, 17
• For HAIR FALL → Recommend products 16, 17
━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━

        RESPONSE FORMAT (JSON only, no markdown):
        {{
          "generalAdvice": "Personalized advice based on user's {user_profile.get('skinType', 'normal')} skin type (2-3 sentences)",
          "products": [
            {{
              "name": "Cetaphil Moisturizing Cream (550ml)",
              "brand": "Cetaphil",
              "category": "Skincare",
              "price": "₹999",
              "ingredients": ["Glycerin", "Panthenol", "Sweet Almond Oil"],
              "pros": ["Perfect for dry skin", "Dermatologist recommended", "Long-lasting hydration"],
              "cons": ["Large bottle may be inconvenient", "Can feel heavy in humid weather"],
              "purchaseLink": "https://www.amazon.in/dp/B003MJG19K",
              "compatibility": 90,
              "applicationTips": "Apply to clean, damp skin twice daily"
            }},
            {{
              "name": "Simple Kind to Skin Hydrating Light Moisturiser (125ml)",
              "brand": "Simple",
              "category": "Skincare",
              "price": "₹385",
              "ingredients": ["Vitamin E", "Vitamin B5", "Bisabolol"],
              "pros": ["Lightweight formula", "Budget-friendly", "Sensitive skin friendly"],
              "cons": ["May not be enough for very dry skin", "Small size"],
              "purchaseLink": "https://www.nykaa.com/simple-kind-to-skin-hydrating-light-moisturiser/p/129162",
              "compatibility": 85,
              "applicationTips": "Use morning and evening after cleansing"
            }}
          ],
          "additionalTips": ["Drink plenty of water for hydration", "Use sunscreen daily", "Avoid harsh soaps"]
        }}

CRITICAL: Copy product names and links EXACTLY from the database above!

PERSONALIZATION REQUIREMENTS:
✓ Match products to user's: {user_profile.get('skinType', 'normal')} skin / {user_profile.get('hairType', 'straight')} hair
✓ Consider budget: {user_profile.get('skincareBudget', 'mid')} range
✓ Avoid allergens: {user_profile.get('allergens', [])}
✓ Match style preference: {user_profile.get('stylePreference', 'casual')}
✓ Consider Indian climate, humidity, and local availability

⚠️ IF YOU'RE UNSURE ABOUT A PRODUCT'S CURRENT AVAILABILITY OR EXACT LINK:
→ Skip that product and recommend another one you're confident about
→ Better to recommend 3 REAL products than 5 potentially fake ones

Respond ONLY with valid JSON. No markdown, no code blocks, no explanations."""

        if image_data:
            prompt += f"\n\nIMAGE CONTEXT: User has uploaded an image for analysis. Please analyze it in your recommendations."
        
        return prompt
    
    def _format_user_profile(self, user_profile: Dict[str, Any]) -> str:
        """
        Format user profile for prompt inclusion
        
        Args:
            user_profile: User preferences dictionary
            
        Returns:
            Formatted profile string
        """
        if not user_profile:
            return "No specific profile information provided."
        
        profile_parts = []
        
        # Skin information
        profile_parts.append(f"• Skin Type: {user_profile.get('skinType', 'Not specified')}")
        profile_parts.append(f"• Primary Skin Concern: {user_profile.get('skinConcern', 'Not specified')}")
        
        # Hair information
        profile_parts.append(f"• Hair Type: {user_profile.get('hairType', 'Not specified')}")
        if user_profile.get('hairTexture'):
            profile_parts.append(f"• Hair Texture: {user_profile['hairTexture']}")
        if user_profile.get('hairConcern'):
            profile_parts.append(f"• Hair Concern: {user_profile['hairConcern']}")
        
        # Style and budget
        profile_parts.append(f"• Style Preference: {user_profile.get('stylePreference', 'Not specified')}")
        profile_parts.append(f"• Skincare Budget: {user_profile.get('skincareBudget', 'mid-range')}")
        profile_parts.append(f"• Fashion Budget: {user_profile.get('fashionBudget', 'mid-range')}")
        
        # Shopping habits
        if user_profile.get('shoppingFrequency'):
            profile_parts.append(f"• Shopping Frequency: {user_profile['shoppingFrequency']}")
        if user_profile.get('occasions'):
            profile_parts.append(f"• Occasions: {user_profile['occasions']}")
        
        # Demographics
        if user_profile.get('ageRange'):
            profile_parts.append(f"• Age Range: {user_profile['ageRange']}")
        
        # Allergies and preferences
        if user_profile.get('allergens'):
            allergens = ', '.join(user_profile['allergens']) if isinstance(user_profile['allergens'], list) else user_profile['allergens']
            profile_parts.append(f"Allergies/Avoid: {allergens}")
        
        if user_profile.get('preferredIngredients'):
            preferred = ', '.join(user_profile['preferredIngredients']) if isinstance(user_profile['preferredIngredients'], list) else user_profile['preferredIngredients']
            profile_parts.append(f"Preferred Ingredients: {preferred}")
        
        return '\n'.join(profile_parts) if profile_parts else "No specific profile information provided."
